Add typing annotation to Perun

tfiedor opened the pull request:
This PR:
  1. Fixes minor issues, and refactor selected parts of codes;
  2. Adds proper typing to all (but few) modules wrt to newest versions of the Python.

tfiedor committed:
Add mypy check

tfiedor committed:
Fix type errors

various: fix wrong names of namedtuples;
various: add ignore to errors caused by static variables;
various: add type hints to certain places.

tfiedor committed:
Add types to clusterizer

tfiedor committed:
Add types to workload generators

workload: change prints to log.info.

tfiedor committed:
Add typing to fuzz evaluation

strucs: rename several functions;
factory: remove the evaluate generic function (problematic).

tfiedor committed:
Extract getting of the gcov version to structs

tfiedor committed:
Extract skipping of coverage testing to different parameter

fuzz/structs: add default values to paths

tfiedor committed:
Add types to rest of testing

tfiedor committed:
Add types to perun.check

check: add types;
structs: extract ModelRecord to dataclass and to structs;
local_statistics: extract to variables;
nonparam_helpers: rename some variables;
utils: extrat Minor/MajorVersion to dataclass.

tfiedor committed:
Fix type issue with Profile vs ProfileInfo

check: make degradation_between_profiles to accept only Profiles.

tfiedor committed:
Fix issue with CallableDetectionMethod type

tfiedor committed:
Extract target model to individual parameter

tfiedor committed:
Fix minor issue with types

tfiedor committed:
Add typing to runners and fix minor issues

*/run.py: add typing;
local_statistics: fix minor issues;
asserts: fix minor issues.

tfiedor committed:
Fix minor issues in complexity

complexity: fix minor issues in types.

tfiedor committed:
Add types to the rest of the runners

tfiedor committed:
Add types to complexity and basic cli

cli: add typing to master commands;
collect/complexity: add typing to complexity module;
other: fix typing for some.

tfiedor committed:
Add types to perunfuzz

perun/fuzz: add typing.

tfiedor committed:
Add typing to memory collector

collect/memory: add typing.

tfiedor committed:
Add types to regression_analysis

regression_analysis: add typing.

tfiedor committed:
Add typing to rest of postprocess

perun/postprocess: add typing.

tfiedor committed:
Add typing to vcs and cli_groups

cli_groups: add typing;
vcs: add typing.

tfiedor committed:
Add typing to perun logic

perun/log: add typing.

tfiedor committed:
Add typing to profile

perun/profile: add typing.

tfiedor committed:
Fix minor issues in typing

tfiedor committed:
Finish adding typing

perun/utils: add typing.

tfiedor committed:
Fix minor issues in typing

tfiedor committed:
Fix minor typing and test issues

tfiedor committed:
Fix minor issues in tests

stats: revert old change;
asserts: fix minor issue.

tfiedor committed:
Fix minor issue in stats

tfiedor committed:
Fix missing return types

tfiedor committed:
Fix minor things in utils

tfiedor committed:
Fix minor issues in typing

tfiedor committed:
Refactor some awkward stuff in fuzzing

perun/fuzz: remove useless functions;
perun/fuzz: switch to formating string.

tfiedor committed:
Fix minor typing issues

tfiedor committed:
Fix typing issues

tfiedor committed:
Fix typing issues

tfiedor committed:
Fix more typing issues

tfiedor committed:
Fix more typing issues

tfiedor committed:
Fix more tiping issues

tfiedor committed:
Fix last batch of typing issues

tfiedor commented:
There is still missing typing of `perun/trace/*` and `perun/thirdparty`. @JiriPavela do you want to fix them yourself?

tfiedor committed:
Fix codacy issues

tfiedor committed:
Fix minor codacy issue

tfiedor committed:
Fix error

tfiedor committed:
Add __future__ typing annotations to modules

tfiedor committed:
Add missing __future__ import

tfiedor committed:
Hotfix python 3.8 issue

tfiedor committed:
Remove support for python 3.8

tfiedor committed:
Remove py38 from github actions

tfiedor committed:
Add coveragerc

tfiedor committed:
Add few tests

commands: add tests for new errors.

tfiedor committed:
Improve coverage

tfiedor committed:
Fix minor things for coverage

various: move coverage info to pyproject.toml;
tests: add duration of tests to `make test`;
test_collect: add few tests.

tfiedor committed:
Add excluded lines

tfiedor committed:
Add ignored typing-related lines

tfiedor committed:
Add tests for logger

tfiedor committed:
Fix unused import

tfiedor committed:
Add tests to metrics

tfiedor committed:
Extract setup to isolate action

tfiedor committed:
Move action where it belongs

tfiedor committed:
Add shells to actions

tfiedor committed:
Add typing and linting to GA

tfiedor committed:
Remove linting

tfiedor committed:
Finish typing initialization

tfiedor committed:
Fix minor issue with Python 3.9

JiriPavela commented on perun/check/factory.py:
```diff
@@ -289,10 +314,10 @@ def parse_strategy(strategy):
     """Retrieves the best strategy for the given profile configuration
 
-    :param ProfileInfo profile: Profile information with configuration tuple
+    :param Profileprofile: Profile information with configuration tuple
```
Seems like a small typo here. However, this leads me to a more general issue: Should we still continue using the old type hints in docstrings now that our codebase is mostly type annotated? This seems a bit redundant to me.

tfiedor replied:
I kept them there, since it seems that some checkers/IDEs/etc might use them (I think PyCharm---though it might be older version---complained with highlight and it annoyed me. But we might stop use them and slowly remove them.

JiriPavela commented on perun/check/integral_comparison.py:
```diff
@@ -19,7 +26,7 @@
 
 
-def compute_param_integral(model):
+def compute_param_integral(model: perun.utils.structs.ModelRecord) -> float:
```
Is there a reason why the ModelRecord uses a fully specified path? Was there perhaps a problem with using just `ModelRecord`?

tfiedor replied:
I think this might be automatic change by PyCharm. I see no reason for full path.

JiriPavela commented on perun/check/linear_regression.py:
```diff
@@ -27,9 +40,14 @@ def linear_regression(baseline_profile, target_profile, **_):
+        lin_abs_error: npt.NDArray[numpy.float64] | list[float],
+        threshold: int,
+        linear_diff_b1: float,
+        baseline_model: perun.utils.structs.ModelRecord, target_model: perun.utils.structs.ModelRecord,
```
Another occurrence of the same thing as in `integral_comparison.py`.

tfiedor replied:
Probaly IDE move.

JiriPavela commented on perun/postprocess/normalizer/run.py:
```diff
@@ -30,7 +33,7 @@ def normalize_resources(resources):
     """
     # First compute maximas per each type
-    maximum_per_type = {}
+    maximum_per_type = {}  # type: dict[str, int]
```
Why not annotate this as `maximum_per_type: dict[str, int] = {}`. Was there some issue with this?

tfiedor replied:
No, old change, when I just started with typing.

JiriPavela commented on perun/postprocess/regression_analysis/data_provider.py:
```diff
@@ -40,8 +47,8 @@ def generic_profile_provider(profile, of_key, per_key, **_):
     resources = sorted(resources, key=resource_sort_key)
-    x_points_list = []
-    y_points_list = []
+    x_points_list = []  # type: list[float]
```
The same thing as in normalizer.

tfiedor replied:
Same as above.

JiriPavela commented on perun/utils/structs.py:
```diff
@@ -300,7 +329,7 @@ def __init__(self, *args):
         self.order = ordered
 
-    def __ge__(self, other):
+    def __ge__(self, other: Any) -> bool:
```
Just a minor remark. I think all the comparison methods are usually typed with `other` being `object` rather than `Any` in the typeshed.

tfiedor replied:
Aggree, `object` is better.

JiriPavela commented on pyproject.toml:
```diff
@@ -163,4 +196,20 @@ omit = [
+
+    # Exclude protocols, that are used for type checking
+    "class CallableDetectionMethod(Protocol):",
+    "class Comparable(Protocol):",
```
This protocol exclusion list seems like a hard-to-maintain thing. Is there perhaps a coverage option to omit type hints related stuff?

tfiedor replied:
I briefly googled, but found no good solution, so, we might add some remark, that this is not best solution. There is not much documentation, I think, regarding the typing and coverage. Naturally, we could write test, which would simply instantiate the class, but it is waste of time.

JiriPavela reviewed (commented):
Good job, let's discuss the best course of action for some of the minor remarks and then merge it.

JiriPavela committed:
Polish type hints.

Fix old style '# type: X' hints, absolute paths where unnecessary
and 'Any' being used instead of 'object' in comparison operators.
